Add wxGLCanvasX11::GLXSetSwapInterval()

vadz opened the pull request:
Allow setting the swap interval value or not setting it at all instead of always forcing it to 0.

The changes of 3dde6bdeb0 (Disable swap interval in GLX wxGLCanvas implementation too, 2023-12-25) are not always desirable, so allow to customize the behaviour.

See #24163, #24165.

vadz committed:
Add wxGLCanvas::SetSwapInterval() and GetSwapInterval()

Allow setting the swap interval value or not setting it at all instead
of always forcing it to 0.

The changes of 3dde6bdeb0 (Disable swap interval in GLX wxGLCanvas
implementation too, 2023-12-25) are not always desirable, so allow to
customize the behaviour.

Implement this for Unix (GLX and EGL) backends for now and allow testing
using these functions in the cube sample.

See #24163, #24165.

dsa-t commented:
I think `SetSwapInterval` needs to be implemented in each GL backend, not just `wxGLCanvasX11::GLXSetSwapInterval`.

And I don't think the interval should be changed inside `SwapBuffers`.

For Wayland, I think a solution similar to SDL can be implemented (see https://github.com/wxWidgets/wxWidgets/issues/24977#issuecomment-2981633368)

vadz committed:
Check for GLX_EXT_swap_control availability

glXSwapIntervalEXT() might exist in the client library but not be
actually supported by the driver, so check if it advertises the
extension defining this function.

Also don't query GLX_SWAP_INTERVAL_EXT if this extension is not
available as it would generate an X error.

vadz committed:
Fall back on GLX_MESA_swap_control if GLX_EXT_swap_control is NA

Try using the older extension if we don't have the newer one.

vadz commented:
> I think `SetSwapInterval` needs to be implemented in each GL backend, not just `wxGLCanvasX11::GLXSetSwapInterval`.

I may have forgotten (or just misunderstood) things but I thought that we wanted to add it to allow customizing setting of the swap interval to 0 done right now in Unix (GLX and EGL) backends. Do we really need it elsewhere? It wouldn't be that difficult to call `wglSwapIntervalEXT()` under Windows and apparently there is `CGLSetParameter(kCGLCPSwapInterval)` for macOS, but would this be really useful there? Will KiCad use this?

> And I don't think the interval should be changed inside `SwapBuffers`.

Note that this is only done once. It's a way to allow setting the interval before it can be applied. Is it still a problem?
 
> For Wayland, I think a solution similar to SDL can be implemented (see [#24977 (comment)](https://github.com/wxWidgets/wxWidgets/issues/24977#issuecomment-2981633368))

This seems complicated and fragile :-( The original commit message speaks about "upcoming protocols that should improve the situation", would anybody know if anything has happened since then (2021)?

> `glXSwapIntervalEXT` function may exist in the client library, but `GLX_EXT_swap_control/GLX_EXT_swap_control_tear` extensions may not supported by the driver.

Thanks, this is definitely a good point, I've updated this PR with a better check. I've also added a check for MESA extension, although I'm unable to test this because all my systems support `EXT_swap_control` — but it looks like it shouldn't do any harm. I didn't add support for `GLX_SGI_swap_control`, even though it would be easy to do, of course, because it looks like this one shouldn't be relevant any longer.

Finally, I've rebased everything on master.

Next steps are:

1. Decide if we need this under Windows (easy to do, questionable utility).
2. And under macOS (slightly less clear how to do, same question about utility).
3. Implement at least `GLXDontSetSwapInterval()` for EGL backend (simple to do, might be useful?).
4. Think about whether we really need `GLXSetSwapInterval()` with arbitrary interval for EGL too and do something like what SDL does, but I'm still not sure if this is really a wise thing to do — and would definitely welcome any better ideas!

vadz commented:
This would need to be redone once #26023 is merged, but I am still not sure about (4) (and also (1) and (2)...) from above.

vadz committed:
Remove unused WXGLSwapBuffers() function

This seems to have never been used in Cocoa wxOSX version and is
definitely not used currently.

vadz committed:
Implement support for wxGLCanvas swap interval in wxOSX

Just use NSOpenGLCPSwapInterval parameter for this.

vadz committed:
Implement support for wxGLCanvas swap interval in wxMSW

Just call the corresponding WGL_EXT_swap_control extension functions if
available.

dsa-t commented:
Thanks!

Adaptive vsync would be useful to have as this is what KiCad currently requests for balance between latency and tearing.

There's also an adaptive vsync implementation for macOS in SDL: https://github.com/libsdl-org/SDL/blob/3dc48a4890d37560e4159f50b7a6eadc9b9c1936/src/video/cocoa/SDL_cocoaopengl.m#L498

I didn't test on MSW yet.

> I think we'll have to leave any smart SDL-like solutions avoiding disabling vsync under Wayland for later, I still don't fully understand what needs to be done and almost certainly won't have time to do it before 3.3.2.

It's the only solution I see for Wayland currently, but it's fine to do later.

vadz committed:
Add support for requesting adaptive VSync to wxGLCanvas

Mimic the low level OpenGL APIs by allowing to pass -1 to
SetSwapInterval() to request adaptive VSync and implement this for GLX
and WGL.

Change return type of SetSwapInterval() to allow returning a value
indicating that enabling adaptive VSync failed but standard VSync could
still be set.

vadz commented:
I've added the possibility to request adaptive VSync now, please let me know if you have any comments.

I've hesitated for a long time about the API to use, but finally kept a simple `int interval` parameter for this function instead of adding `int flags` and passing some `Adaptive` flag to it because it looks like using -1/0/1 is standard practice in OpenGL programming and so it probably doesn't make sense to deviate from it.

I still changed the return type to be a bit more type-safe. The alternative here would be to just return `false` if adaptive VSync is not supported, but this would force all applications to do

```cpp
if ( !SetSwapInterval(-1) )
    SetSwapInterval(1);
```

as I don't think there are any cases where you'd want to use adaptive VSync but not want to enable standard VSync if it's not available, so it seems better to do it inside the library.

Finally, please note that I believe that the code in KiCad `GL_UTILS::SetSwapInterval()` is wrong as it queries `GLX_LATE_SWAPS_TEAR_EXT` before requesting adaptive VSync and AFAICS it's always going to be 0 in this case. But please let me know if I'm missing something, of course.

vadz commented:
P.S. Testing under MSW would still be much appreciated!

dsa-t commented:
> Finally, please note that I believe that the code in KiCad `GL_UTILS::SetSwapInterval()` is wrong as it queries `GLX_LATE_SWAPS_TEAR_EXT` before requesting adaptive VSync and AFAICS it's always going to be 0 in this case. But please let me know if I'm missing something, of course.

Looks like Mesa errors out if `EXT_swap_control_tear_bit` is not set at `glXSwapIntervalEXT` call: https://gitlab.freedesktop.org/mesa/mesa/-/blob/51d4803a6fbae1d280885ed6b70683efadd12e81/src/glx/glxcmds.c#L1674

I presume these bits are set internally by mesa and not requested by the user.

dsa-t commented:
Which GPUs and drivers were you testing on?

dsa-t commented:
Anyways I think you're right. 

Since the extension check was added, the drawable check is not needed.

dsa-t commented:
Apparently the "GDI Generic", version "1.1.0" driver is used at `wxGLContext::wxGLContext`.
But when running KiCad (instead of wx samples), it recognizes the NVIDIA driver correctly.

vadz commented:
> API looks good for me. In KiCad it would be used like this:
> 
> ```
>     wxGLCanvas::SetSwapInterval( -1 );
>     m_swapInterval = wxGLCanvas::GetSwapInterval();
> ```

Just to understand, why do you need the actual swap interval back? How do you use it and will all (or at least most) applications need it to? If so, maybe it would make sense to return it from `SetSwapInterval()`, after all... I just thought it would be something most applications wouldn't really care about because what can you do if you've failed to set adaptive VSync, there doesn't seem to be any reasonable fallback other than using standard VSync, is there?

> On Win11 (NVIDIA GPU), `wglGetProcAddress("wglGetExtensionsStringARB")` in `wxGLCanvasBase::IsExtensionSupported` returns null. As well as other `wglGetProcAddress` calls in `wxGLContext::wxGLContext`. So the context is created via "legacy" `wglCreateContext`.

In my VM this happens because it's using the generic GDI driver (as also confirmed by running `wglinfo`, i.e. it's not something wx-specific), but I hoped it wouldn't be the case in a physical machine (or maybe better configured VMs, as I don't care about graphics much I've never really looked into this).

>  But when running KiCad (instead of wx samples), it recognizes the NVIDIA driver correctly.

So in your case `wglinfo` sees Nvidia but wx cube sample still shows "GDI Generic" in its "About"? This would be a bug in wx...
